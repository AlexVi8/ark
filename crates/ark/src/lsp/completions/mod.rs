//
// mod.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

mod completion_item;
mod document;
mod provide;
mod resolve;
mod session;
mod types;
mod workspace;

pub use provide::provide_completions;
pub use resolve::resolve_completion;
