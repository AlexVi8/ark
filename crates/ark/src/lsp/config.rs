use serde::Deserialize;
use serde::Serialize;
use serde_json::Value;

use crate::lsp::diagnostics::DiagnosticsConfig;

pub struct Setting<T> {
    pub key: &'static str,
    pub set: fn(&mut T, Value),
}

/// List of LSP settings for which clients can send `didChangeConfiguration`
/// notifications. We register our interest in watching over these settings in
/// our `initialized` handler. The `set` methods convert from a json `Value` to
/// the expected type, using a default value if the conversion fails.
///
/// This array is for global settings. If the setting should only affect a given
/// document URI, add it to `DOCUMENT_SETTINGS` instead.
pub static GLOBAL_SETTINGS: &[Setting<LspConfig>] = &[
    Setting {
        key: "positron.r.diagnostics.enable",
        set: |cfg, v| {
            cfg.diagnostics.enable = v
                .as_bool()
                .unwrap_or_else(|| DiagnosticsConfig::default().enable)
        },
    },
    Setting {
        key: "positron.r.symbols.includeAssignmentsInBlocks",
        set: |cfg, v| {
            cfg.symbols.include_assignments_in_blocks = v
                .as_bool()
                .unwrap_or_else(|| SymbolsConfig::default().include_assignments_in_blocks)
        },
    },
    Setting {
        key: "positron.r.workspaceSymbols.includeCommentSections",
        set: |cfg, v| {
            cfg.workspace_symbols.include_comment_sections = v
                .as_bool()
                .unwrap_or_else(|| WorkspaceSymbolsConfig::default().include_comment_sections)
        },
    },
];

/// These document settings are updated on a URI basis. Each document has its
/// own value of the setting.
pub static DOCUMENT_SETTINGS: &[Setting<DocumentConfig>] = &[
    Setting {
        key: "editor.insertSpaces",
        set: |cfg, v| {
            let default_style = IndentationConfig::default().indent_style;
            cfg.indent.indent_style = if v
                .as_bool()
                .unwrap_or_else(|| default_style == IndentStyle::Space)
            {
                IndentStyle::Space
            } else {
                IndentStyle::Tab
            }
        },
    },
    Setting {
        key: "editor.indentSize",
        set: |cfg, v| {
            cfg.indent.indent_size = v
                .as_u64()
                .map(|n| n as usize)
                .unwrap_or_else(|| IndentationConfig::default().indent_size)
        },
    },
    Setting {
        key: "editor.tabSize",
        set: |cfg, v| {
            cfg.indent.tab_width = v
                .as_u64()
                .map(|n| n as usize)
                .unwrap_or_else(|| IndentationConfig::default().tab_width)
        },
    },
];

/// Configuration of the LSP
#[derive(Clone, Default, Debug)]
pub(crate) struct LspConfig {
    pub(crate) diagnostics: DiagnosticsConfig,
    pub(crate) symbols: SymbolsConfig,
    pub(crate) workspace_symbols: WorkspaceSymbolsConfig,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct SymbolsConfig {
    /// Whether to emit assignments in `{` bloks as document symbols.
    pub include_assignments_in_blocks: bool,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct WorkspaceSymbolsConfig {
    /// Whether to include sections like `# My section ---` in workspace symbols.
    pub include_comment_sections: bool,
}

/// Configuration of a document.
///
/// The naming follows <https://editorconfig.org/> where possible.
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct DocumentConfig {
    pub indent: IndentationConfig,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct IndentationConfig {
    /// Whether to insert spaces of tabs for one level of indentation.
    pub indent_style: IndentStyle,

    /// The number of spaces for one level of indentation.
    pub indent_size: usize,

    /// The width of a tab. There may be projects with an `indent_size` of 4 and
    /// a `tab_width` of 8 (e.g. GNU R).
    pub tab_width: usize,
}

#[derive(PartialEq, Serialize, Deserialize, Clone, Debug)]
pub enum IndentStyle {
    Tab,
    Space,
}

impl Default for SymbolsConfig {
    fn default() -> Self {
        Self {
            include_assignments_in_blocks: false,
        }
    }
}

impl Default for WorkspaceSymbolsConfig {
    fn default() -> Self {
        Self {
            include_comment_sections: false,
        }
    }
}

impl Default for IndentationConfig {
    fn default() -> Self {
        Self {
            indent_style: IndentStyle::Space,
            indent_size: 2,
            tab_width: 2,
        }
    }
}

pub(crate) fn indent_style_from_lsp(insert_spaces: bool) -> IndentStyle {
    if insert_spaces {
        IndentStyle::Space
    } else {
        IndentStyle::Tab
    }
}
