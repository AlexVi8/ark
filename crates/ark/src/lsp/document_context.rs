//
// document_context.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use tree_sitter::Node;
use tree_sitter::Point;

use crate::lsp::documents::Document;
use crate::lsp::traits::node::NodeExt;

#[derive(Debug)]
pub struct DocumentContext<'a> {
    pub document: &'a Document,
    pub node: Node<'a>,
    pub source: String,
    pub point: Point,
}

impl<'a> DocumentContext<'a> {
    pub fn new(document: &'a Document, point: Point) -> Self {
        // get reference to AST
        let ast = &document.ast;

        // find node at point
        let node = ast.root_node().find_closest_node_to_point(point).unwrap();

        // convert document contents to a string once, to be reused elsewhere
        let source = document.contents.to_string();

        // build document context
        DocumentContext {
            document,
            node,
            source,
            point,
        }
    }
}
