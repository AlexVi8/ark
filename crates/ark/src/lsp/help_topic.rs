//
// help_topic.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use serde::Deserialize;
use serde::Serialize;
use tower_lsp::lsp_types::Position;
use tower_lsp::lsp_types::VersionedTextDocumentIdentifier;

use crate::backend_trace;
use crate::lsp::backend::Backend;
use crate::lsp::traits::cursor::TreeCursorExt;
use crate::lsp::traits::position::PositionExt;

pub static POSITRON_HELP_TOPIC_REQUEST: &'static str = "positron/textDocument/helpTopic";

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HelpTopicParams {
    /// The document to provide a help topic for.
    pub text_document: VersionedTextDocumentIdentifier,
    /// The location of the cursor.
    pub position: Position,
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HelpTopicResponse {
    /// The help topic appropriate for the cursor position.
    pub topic: String,
}

impl Backend {
    pub async fn help_topic(
        &self,
        params: HelpTopicParams,
    ) -> tower_lsp::jsonrpc::Result<Option<HelpTopicResponse>> {
        backend_trace!(self, "help_topic({:?})", params);

        let uri = &params.text_document.uri;
        let Some(document) = self.documents.get_mut(uri) else {
            backend_trace!(self, "help_topic(): No document associated with URI {uri}");
            return Ok(None);
        };

        let root = document.ast.root_node();
        let position = params.position;
        let point = position.as_point();

        let mut cursor = root.walk();
        let node = cursor.find_leaf(point);

        // Check to see whether it's a function node (currently that's what we
        // support)
        if node.kind() != "function" {
            backend_trace!(self, "help_topic(): No function node at position {point}");
            return Ok(None);
        }

        // Get the text of the node
        let source = document.contents.to_string();
        let text = node.utf8_text(source.as_bytes()).unwrap();

        let response = HelpTopicResponse {
            topic: String::from(text),
        };

        Ok(Some(response))
    }
}
