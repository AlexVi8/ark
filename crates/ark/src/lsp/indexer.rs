//
// indexer.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

use std::collections::HashMap;
use std::path::Path;
use std::result::Result::Ok;
use std::sync::Arc;
use std::sync::LazyLock;
use std::sync::Mutex;

use anyhow::anyhow;
use regex::Regex;
use ropey::Rope;
use stdext::unwrap;
use stdext::unwrap::IntoResult;
use tower_lsp::lsp_types::Range;
use tree_sitter::Node;
use walkdir::DirEntry;
use walkdir::WalkDir;

use crate::lsp;
use crate::lsp::documents::Document;
use crate::lsp::encoding::convert_point_to_position;
use crate::lsp::traits::rope::RopeExt;
use crate::treesitter::BinaryOperatorType;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;

#[derive(Clone, Debug)]
pub enum IndexEntryData {
    Variable {
        name: String,
    },
    Function {
        name: String,
        arguments: Vec<String>,
    },
    // Like Function but not used for completions yet
    Method {
        name: String,
    },
    Section {
        level: usize,
        title: String,
    },
}

#[derive(Clone, Debug)]
pub struct IndexEntry {
    pub key: String,
    pub range: Range,
    pub data: IndexEntryData,
}

type DocumentPath = String;
type DocumentSymbol = String;
type DocumentSymbolIndex = HashMap<DocumentSymbol, IndexEntry>;
type WorkspaceIndex = Arc<Mutex<HashMap<DocumentPath, DocumentSymbolIndex>>>;

static WORKSPACE_INDEX: LazyLock<WorkspaceIndex> = LazyLock::new(|| Default::default());
pub static RE_COMMENT_SECTION: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\s*(#+)\s*(.*?)\s*[#=-]{4,}\s*$").unwrap());

#[tracing::instrument(level = "info", skip_all)]
pub fn start(folders: Vec<String>) {
    let now = std::time::Instant::now();
    lsp::log_info!("Initial indexing started");

    for folder in folders {
        let walker = WalkDir::new(folder);
        for entry in walker.into_iter().filter_entry(|e| filter_entry(e)) {
            if let Ok(entry) = entry {
                if entry.file_type().is_file() {
                    if let Err(err) = index_file(entry.path()) {
                        lsp::log_error!("Can't index file {:?}: {err:?}", entry.path());
                    }
                }
            }
        }
    }

    lsp::log_info!(
        "Initial indexing finished after {}ms",
        now.elapsed().as_millis()
    );
}

pub fn find(symbol: &str) -> Option<(String, IndexEntry)> {
    let index = WORKSPACE_INDEX.lock().unwrap();

    for (path, index) in index.iter() {
        if let Some(entry) = index.get(symbol) {
            return Some((path.clone(), entry.clone()));
        }
    }

    None
}

pub fn map(mut callback: impl FnMut(&Path, &String, &IndexEntry)) {
    let index = WORKSPACE_INDEX.lock().unwrap();

    for (path, index) in index.iter() {
        for (symbol, entry) in index.iter() {
            let path = Path::new(path);
            callback(path, symbol, entry);
        }
    }
}

#[tracing::instrument(level = "trace", skip_all, fields(path = ?path))]
pub fn update(document: &Document, path: &Path) -> anyhow::Result<()> {
    clear(path)?;
    index_document(document, path);
    Ok(())
}

fn insert(path: &Path, entry: IndexEntry) -> anyhow::Result<()> {
    let mut index = WORKSPACE_INDEX.lock().unwrap();
    let path = str_from_path(path)?;

    let index = index.entry(path.to_string()).or_default();

    // Retain the first occurrence in the index. In the future we'll track every occurrences and
    // their scopes but for now we only track the first definition of an object (in a way, its
    // declaration).
    if !index.contains_key(&entry.key) {
        index.insert(entry.key.clone(), entry);
    }

    Ok(())
}

fn clear(path: &Path) -> anyhow::Result<()> {
    let mut index = WORKSPACE_INDEX.lock().unwrap();
    let path = str_from_path(path)?;

    // Only clears if the `path` was an existing key
    index.entry(path.into()).and_modify(|index| {
        index.clear();
    });

    Ok(())
}

fn str_from_path(path: &Path) -> anyhow::Result<&str> {
    path.to_str().ok_or(anyhow!(
        "Couldn't convert path {} to string",
        path.display()
    ))
}

// TODO: Should we consult the project .gitignore for ignored files?
// TODO: What about front-end ignores?
// TODO: What about other kinds of ignores (e.g. revdepcheck)?
pub fn filter_entry(entry: &DirEntry) -> bool {
    let name = entry.file_name();

    // skip common ignores
    for ignore in [".git", ".Rproj.user", "node_modules", "revdep"] {
        if name == ignore {
            return false;
        }
    }

    // skip project 'renv' folder
    if name == "renv" {
        let companion = entry.path().join("activate.R");
        if companion.exists() {
            return false;
        }
    }

    true
}

fn index_file(path: &Path) -> anyhow::Result<()> {
    // only index R files
    let ext = path.extension().unwrap_or_default();
    if ext != "r" && ext != "R" {
        return Ok(());
    }

    // TODO: Handle document encodings here.
    // TODO: Check if there's an up-to-date buffer to be used.
    let contents = std::fs::read(path)?;
    let contents = String::from_utf8(contents)?;
    let document = Document::new(contents.as_str(), None);

    index_document(&document, path);

    Ok(())
}

fn index_document(document: &Document, path: &Path) {
    let ast = &document.ast;
    let contents = &document.contents;

    let root = ast.root_node();
    let mut cursor = root.walk();
    let mut entries = Vec::new();

    for node in root.children(&mut cursor) {
        if let Err(err) = index_node(path, contents, &node, &mut entries) {
            lsp::log_error!("Can't index document: {err:?}");
        }
    }

    for entry in entries {
        if let Err(err) = insert(path, entry) {
            lsp::log_error!("Can't insert index entry: {err:?}");
        }
    }
}

fn index_node(
    path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    index_assignment(path, contents, node, entries)?;
    index_comment(path, contents, node, entries)?;
    Ok(())
}

fn index_assignment(
    path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    if !matches!(
        node.node_type(),
        NodeType::BinaryOperator(BinaryOperatorType::LeftAssignment) |
            NodeType::BinaryOperator(BinaryOperatorType::EqualsAssignment)
    ) {
        return Ok(());
    }

    let lhs = match node.child_by_field_name("lhs") {
        Some(lhs) => lhs,
        None => return Ok(()),
    };

    let Some(rhs) = node.child_by_field_name("rhs") else {
        return Ok(());
    };

    if crate::treesitter::node_is_call(&rhs, "R6Class", contents) ||
        crate::treesitter::node_is_namespaced_call(&rhs, "R6", "R6Class", contents)
    {
        index_r6_class(path, contents, &rhs, entries)?;
    }

    let lhs_text = contents.node_slice(&lhs)?.to_string();

    // The method matching is super hacky but let's wait until the typed API to
    // do better
    if !lhs_text.starts_with("method(") && !lhs.is_identifier_or_string() {
        return Ok(());
    }

    let Some(rhs) = node.child_by_field_name("rhs") else {
        return Ok(());
    };

    if rhs.is_function_definition() {
        // If RHS is a function definition, emit a function symbol
        let mut arguments = Vec::new();
        if let Some(parameters) = rhs.child_by_field_name("parameters") {
            let mut cursor = parameters.walk();
            for child in parameters.children(&mut cursor) {
                let name = unwrap!(child.child_by_field_name("name"), None => continue);
                if name.is_identifier() {
                    let name = contents.node_slice(&name)?.to_string();
                    arguments.push(name);
                }
            }
        }

        let start = convert_point_to_position(contents, lhs.start_position());
        let end = convert_point_to_position(contents, lhs.end_position());

        entries.push(IndexEntry {
            key: lhs_text.clone(),
            range: Range { start, end },
            data: IndexEntryData::Function {
                name: lhs_text,
                arguments,
            },
        });
    } else {
        // Otherwise, emit variable
        let start = convert_point_to_position(contents, lhs.start_position());
        let end = convert_point_to_position(contents, lhs.end_position());
        entries.push(IndexEntry {
            key: lhs_text.clone(),
            range: Range { start, end },
            data: IndexEntryData::Variable { name: lhs_text },
        });
    }

    Ok(())
}

fn index_r6_class(
    _path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    let Some(args_node) = node.child_by_field_name("arguments") else {
        return Ok(());
    };

    let mut cursor = args_node.walk();
    for arg in args_node.children(&mut cursor) {
        // Only consider `public = ` and `private = ` arguments
        let Some(arg_name) = arg.child_by_field_name("name") else {
            continue;
        };
        if !arg_name.is_identifier_or_string() {
            continue;
        }
        let arg_name_str = contents.node_slice(&arg_name)?;
        if arg_name_str != "public" && arg_name_str != "private" {
            continue;
        }

        let Some(list_node) = arg.child_by_field_name("value") else {
            continue;
        };
        if !list_node.is_call() {
            continue;
        }

        let Some(list_args) = list_node.child_by_field_name("arguments") else {
            return Ok(());
        };

        let mut cursor = list_args.walk();
        for arg in list_args.children(&mut cursor) {
            if !arg.is_argument() {
                continue;
            }

            let (Some(mtd_name), Some(mtd_value)) = (
                arg.child_by_field_name("name"),
                arg.child_by_field_name("value"),
            ) else {
                continue;
            };
            if !mtd_name.is_identifier_or_string() || !mtd_value.is_function_definition() {
                continue;
            }

            let name = contents.node_slice(&mtd_name)?.to_string();
            let start = convert_point_to_position(contents, mtd_name.start_position());
            let end = convert_point_to_position(contents, mtd_name.end_position());

            entries.push(IndexEntry {
                key: name.clone(),
                range: Range { start, end },
                data: IndexEntryData::Method { name },
            });
        }
    }

    Ok(())
}

fn index_comment(
    _path: &Path,
    contents: &Rope,
    node: &Node,
    entries: &mut Vec<IndexEntry>,
) -> anyhow::Result<()> {
    // check for comment
    if !node.is_comment() {
        return Ok(());
    }

    // see if it looks like a section
    let comment = contents.node_slice(node)?.to_string();
    let matches = match RE_COMMENT_SECTION.captures(comment.as_str()) {
        Some(m) => m,
        None => return Ok(()),
    };

    let level = matches.get(1).into_result()?;
    let title = matches.get(2).into_result()?;

    let level = level.as_str().len();
    let title = title.as_str().to_string();

    // skip things that look like knitr output
    if title.starts_with("----") {
        return Ok(());
    }

    let start = convert_point_to_position(contents, node.start_position());
    let end = convert_point_to_position(contents, node.end_position());

    entries.push(IndexEntry {
        key: title.clone(),
        range: Range::new(start, end),
        data: IndexEntryData::Section { level, title },
    });

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::assert_debug_snapshot;

    use super::*;
    use crate::lsp::documents::Document;

    macro_rules! test_index {
        ($code:expr) => {
            let doc = Document::new($code, None);
            let path = PathBuf::from("/path/to/file.R");
            let root = doc.ast.root_node();
            let mut cursor = root.walk();

            let mut entries = vec![];
            for node in root.children(&mut cursor) {
                let _ = index_node(&path, &doc.contents, &node, &mut entries);
            }
            assert_debug_snapshot!(entries);
        };
    }

    // Note that unlike document symbols whose ranges cover the whole entity
    // they represent, the range of workspace symbols only cover the identifers

    #[test]
    fn test_index_function() {
        test_index!(
            r#"
my_function <- function(a, b = 1) {
  a + b

  # These are not indexed as workspace symbol
  inner <- function() {
    2
  }
  inner_var <- 3
}

my_variable <- 1
"#
        );
    }

    #[test]
    fn test_index_variable() {
        test_index!(
            r#"
x <- 10
y = "hello"
"#
        );
    }

    #[test]
    fn test_index_s7_methods() {
        test_index!(
            r#"
Class <- new_class("Class")
generic <- new_generic("generic", "arg",
  function(arg) {
    S7_dispatch()
  }
)
method(generic, Class) <- function(arg) {
  NULL
}
"#
        );
    }

    #[test]
    fn test_index_comment_section() {
        test_index!(
            r#"
# Section 1 ----
x <- 10

## Subsection ======
y <- 20

x <- function() {
    # This inner section is not indexed ----
}

"#
        );
    }

    #[test]
    fn test_index_r6class() {
        test_index!(
            r#"
class <- R6Class(
    public = list(
        initialize = function() {
            1
        },
        public_method = function() {
            2
        }
    ),
    private = list(
        private_method = function() {
            1
        }
    ),
    other = list(
        other_method = function() {
            1
        }
    )
)
"#
        );
    }

    #[test]
    fn test_index_r6class_namespaced() {
        test_index!(
            r#"
class <- R6::R6Class(
    public = list(
        initialize = function() {
            1
        },
    )
)
"#
        );
    }
}
