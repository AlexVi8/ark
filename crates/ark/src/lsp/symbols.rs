//
// symbols.rs
//
// Copyright (C) 2022-2024 Posit Software, PBC. All rights reserved.
//
//

#![allow(deprecated)]

use std::result::Result::Ok;

use ropey::Rope;
use stdext::unwrap::IntoResult;
use tower_lsp::lsp_types::DocumentSymbol;
use tower_lsp::lsp_types::DocumentSymbolParams;
use tower_lsp::lsp_types::Location;
use tower_lsp::lsp_types::Range;
use tower_lsp::lsp_types::SymbolInformation;
use tower_lsp::lsp_types::SymbolKind;
use tower_lsp::lsp_types::Url;
use tower_lsp::lsp_types::WorkspaceSymbolParams;
use tree_sitter::Node;

use crate::lsp::encoding::convert_point_to_position;
use crate::lsp::indexer;
use crate::lsp::indexer::IndexEntryData;
use crate::lsp::state::WorldState;
use crate::lsp::traits::rope::RopeExt;
use crate::lsp::traits::string::StringExt;
use crate::treesitter::BinaryOperatorType;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;

fn new_symbol_node(
    name: String,
    kind: SymbolKind,
    detail: Option<String>,
    range: Range,
    children: Vec<DocumentSymbol>,
) -> DocumentSymbol {
    let mut symbol = new_symbol(name, kind, detail, range);
    symbol.children = Some(children);
    symbol
}

fn new_symbol(
    name: String,
    kind: SymbolKind,
    detail: Option<String>,
    range: Range,
) -> DocumentSymbol {
    DocumentSymbol {
        name,
        kind,
        detail,
        // Safety: We assume `children` can't be `None`
        children: Some(Vec::new()),
        deprecated: None,
        tags: None,
        range,
        selection_range: range,
    }
}

pub fn symbols(params: &WorkspaceSymbolParams) -> anyhow::Result<Vec<SymbolInformation>> {
    let query = &params.query;
    let mut info: Vec<SymbolInformation> = Vec::new();

    indexer::map(|path, symbol, entry| {
        if !symbol.fuzzy_matches(query) {
            return;
        }

        match &entry.data {
            IndexEntryData::Function { name, arguments: _ } => {
                info.push(SymbolInformation {
                    name: name.to_string(),
                    kind: SymbolKind::FUNCTION,
                    location: Location {
                        uri: Url::from_file_path(path).unwrap(),
                        range: entry.range,
                    },
                    tags: None,
                    deprecated: None,
                    container_name: None,
                });
            },

            IndexEntryData::Section { level: _, title } => {
                info.push(SymbolInformation {
                    name: title.to_string(),
                    kind: SymbolKind::STRING,
                    location: Location {
                        uri: Url::from_file_path(path).unwrap(),
                        range: entry.range,
                    },
                    tags: None,
                    deprecated: None,
                    container_name: None,
                });
            },
        };
    });

    Ok(info)
}

pub(crate) fn document_symbols(
    state: &WorldState,
    params: &DocumentSymbolParams,
) -> anyhow::Result<Vec<DocumentSymbol>> {
    let uri = &params.text_document.uri;
    let document = state.documents.get(uri).into_result()?;
    let ast = &document.ast;
    let contents = &document.contents;

    let node = ast.root_node();

    // Index from the root
    match index_node(&node, vec![], &contents) {
        Ok(children) => Ok(children),
        Err(err) => {
            log::error!("Error indexing node: {err:?}");
            return Ok(Vec::new());
        },
    }
}

fn index_node(
    node: &Node,
    mut store: Vec<DocumentSymbol>,
    contents: &Rope,
) -> anyhow::Result<Vec<DocumentSymbol>> {
    // Check if the node is a comment and matches the markdown-style comment patterns
    if node.node_type() == NodeType::Comment {
        let comment_text = contents.node_slice(&node)?.to_string();

        // Check if the comment starts with one or more '#' followed by any text and ends with 4+ punctuations
        if let Some((_level, title)) = parse_comment_as_section(&comment_text) {
            // Create a symbol based on the parsed comment
            let start = convert_point_to_position(contents, node.start_position());
            let end = convert_point_to_position(contents, node.end_position());

            let symbol = new_symbol(title, SymbolKind::STRING, None, Range { start, end });
            store.push(symbol);

            // Return early to avoid further processing
            return Ok(store);
        }
    }

    if matches!(
        node.node_type(),
        NodeType::BinaryOperator(BinaryOperatorType::LeftAssignment) |
            NodeType::BinaryOperator(BinaryOperatorType::EqualsAssignment)
    ) {
        return index_assignment(node, store, contents);
    }

    // Recurse into children. We're in the same outline section so use the same store.
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        if is_indexable(&child) {
            store = index_node(&child, store, contents)?;
        }
    }

    Ok(store)
}

fn index_assignment(
    node: &Node,
    mut store: Vec<DocumentSymbol>,
    contents: &Rope,
) -> anyhow::Result<Vec<DocumentSymbol>> {
    // Check for assignment
    matches!(
        node.node_type(),
        NodeType::BinaryOperator(BinaryOperatorType::LeftAssignment) |
            NodeType::BinaryOperator(BinaryOperatorType::EqualsAssignment)
    )
    .into_result()?;

    // check for lhs, rhs
    let lhs = node.child_by_field_name("lhs").into_result()?;
    let rhs = node.child_by_field_name("rhs").into_result()?;

    // check for identifier on lhs, function on rhs
    let function = lhs.is_identifier_or_string() && rhs.is_function_definition();

    if function {
        return index_assignment_with_function(node, store, contents);
    }

    // otherwise, just index as generic object
    let name = contents.node_slice(&lhs)?.to_string();

    let start = convert_point_to_position(contents, lhs.start_position());
    let end = convert_point_to_position(contents, lhs.end_position());

    let symbol = new_symbol(name, SymbolKind::VARIABLE, None, Range { start, end });
    store.push(symbol);

    Ok(store)
}

fn index_assignment_with_function(
    node: &Node,
    mut store: Vec<DocumentSymbol>,
    contents: &Rope,
) -> anyhow::Result<Vec<DocumentSymbol>> {
    // check for lhs, rhs
    let lhs = node.child_by_field_name("lhs").into_result()?;
    let rhs = node.child_by_field_name("rhs").into_result()?;

    // start extracting the argument names
    let mut arguments: Vec<String> = Vec::new();
    let parameters = rhs.child_by_field_name("parameters").into_result()?;

    let mut cursor = parameters.walk();
    for parameter in parameters.children_by_field_name("parameter", &mut cursor) {
        let name = parameter.child_by_field_name("name").into_result()?;
        let name = contents.node_slice(&name)?.to_string();
        arguments.push(name);
    }

    let name = contents.node_slice(&lhs)?.to_string();
    let detail = format!("function({})", arguments.join(", "));

    let range = Range {
        start: convert_point_to_position(contents, lhs.start_position()),
        end: convert_point_to_position(contents, rhs.end_position()),
    };

    // At this point we increase the nesting level. Recurse into the function
    // node with a new store of children nodes.
    let children = index_node(&rhs, vec![], contents)?;

    let symbol = new_symbol_node(name, SymbolKind::FUNCTION, Some(detail), range, children);
    store.push(symbol);

    Ok(store)
}

fn is_indexable(node: &Node) -> bool {
    // Don't index 'arguments' or 'parameters'
    if matches!(node.node_type(), NodeType::Arguments | NodeType::Parameters) {
        return false;
    }

    true
}

// Function to parse a comment and return the section level and title
fn parse_comment_as_section(comment: &str) -> Option<(usize, String)> {
    // Match lines starting with one or more '#' followed by some non-empty content and must end with 4 or more '-', '#', or `=`
    // Ensure that there's actual content between the start and the trailing symbols.
    if let Some(caps) = indexer::RE_COMMENT_SECTION.captures(comment) {
        let hashes = caps.get(1)?.as_str().len(); // Count the number of '#'
        let title = caps.get(2)?.as_str().trim().to_string(); // Extract the title text without trailing punctuations
        if title.is_empty() {
            return None; // Return None for lines with only hashtags
        }
        return Some((hashes, title)); // Return the level based on the number of '#' and the title
    }

    None
}

#[cfg(test)]
mod tests {
    use tower_lsp::lsp_types::Position;

    use super::*;
    use crate::lsp::documents::Document;

    fn test_symbol(code: &str) -> Vec<DocumentSymbol> {
        let doc = Document::new(code, None);
        let node = doc.ast.root_node();

        index_node(&node, vec![], &doc.contents).unwrap()
    }

    #[test]
    fn test_symbol_parse_comment_as_section() {
        assert_eq!(parse_comment_as_section("# foo"), None);
        assert_eq!(parse_comment_as_section("# foo ---"), None);
        assert_eq!(parse_comment_as_section("########"), None);
        assert_eq!(
            parse_comment_as_section("# foo ----"),
            Some((1, String::from("foo")))
        );
    }

    #[test]
    fn test_symbol_comment_sections() {
        assert_eq!(test_symbol("# foo"), vec![]);
        assert_eq!(test_symbol("# foo ---"), vec![]);
        assert_eq!(test_symbol("########"), vec![]);

        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 10,
            },
        };
        assert_eq!(test_symbol("# foo ----"), vec![new_symbol(
            String::from("foo"),
            SymbolKind::STRING,
            None,
            range
        )]);
    }

    #[test]
    fn test_symbol_assignment() {
        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 3,
            },
        };
        assert_eq!(test_symbol("foo <- 1"), vec![new_symbol(
            String::from("foo"),
            SymbolKind::OBJECT,
            None,
            range,
        )]);
    }

    #[test]
    fn test_symbol_assignment_function() {
        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 20,
            },
        };
        assert_eq!(test_symbol("foo <- function() {}"), vec![new_symbol(
            String::from("foo"),
            SymbolKind::FUNCTION,
            Some(String::from("function()")),
            range,
        )]);
    }

    #[test]
    fn test_symbol_assignment_function_nested() {
        let range = Range {
            start: Position {
                line: 0,
                character: 20,
            },
            end: Position {
                line: 0,
                character: 23,
            },
        };
        let bar = new_symbol(String::from("bar"), SymbolKind::OBJECT, None, range);

        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 30,
            },
        };
        let mut foo = new_symbol(
            String::from("foo"),
            SymbolKind::FUNCTION,
            Some(String::from("function()")),
            range,
        );
        foo.children = Some(vec![bar]);

        assert_eq!(test_symbol("foo <- function() { bar <- 1 }"), vec![foo]);
    }
}
