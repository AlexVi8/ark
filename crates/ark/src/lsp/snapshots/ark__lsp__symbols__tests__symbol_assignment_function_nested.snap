---
source: crates/ark/src/lsp/symbols.rs
expression: "test_symbol(\"foo <- function() { bar <- function() 1 }\")"
---
[
    DocumentSymbol {
        name: "foo",
        detail: Some(
            "function()",
        ),
        kind: Function,
        tags: None,
        deprecated: None,
        range: Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 41,
            },
        },
        selection_range: Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 41,
            },
        },
        children: Some(
            [
                DocumentSymbol {
                    name: "bar",
                    detail: Some(
                        "function()",
                    ),
                    kind: Function,
                    tags: None,
                    deprecated: None,
                    range: Range {
                        start: Position {
                            line: 0,
                            character: 20,
                        },
                        end: Position {
                            line: 0,
                            character: 39,
                        },
                    },
                    selection_range: Range {
                        start: Position {
                            line: 0,
                            character: 20,
                        },
                        end: Position {
                            line: 0,
                            character: 39,
                        },
                    },
                    children: Some(
                        [],
                    ),
                },
            ],
        ),
    },
]
